WIP Bulk update with celery

potsui opened the pull request:
@stvstnfrd @caseylitton @caesar2164

Progress check on bulk update currently running with celery.

TODO
---
- [ ] Finish unit tests
- [ ] Move files to `openedx/stanford` and remove dependencies on `cms`
- [ ] Test system and quality
- [ ] Documentation
- [ ] Test in all browsers
- [ ] Conversion to non-celery as top commit on branch

CC @jspayd @jlikhuva

caseylitton commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+from contentstore.utils import reverse_course_url
+
+
+SHOW_ANSWER_OPTIONS = [
```
Can you import HOW_ANSWER_OPTIONS from  cms/djangoapps/contentstore/views/utilities/bulkupdate.py?

potsui replied:
done

stvstnfrd commented on cms/templates/emails/utilities_bulk_update_done_subject.txt:
```diff
@@ -0,0 +1,4 @@
+<%! from django.utils.translation import ugettext as _ %>
+${_("{username}'s bulk update on {course}").format(
+    username=user_username, course=course_name,
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-multiline.markdown

potsui replied:
done

stvstnfrd commented on cms/templates/emails/utilities_bulk_update_done_message.txt:
```diff
@@ -0,0 +1,14 @@
+<%! from django.utils.translation import ugettext as _ %>
+${_("Hello, {user}").format(user=user_username)}
```
Can we just make this `username=username`?

potsui replied:
done

stvstnfrd commented on cms/templates/emails/utilities_bulk_update_done_message.txt:
```diff
@@ -0,0 +1,14 @@
+${_("Hello, {user}").format(user=user_username)}
+
+% if success:
+${_("Your bulk update on the {course} course has successfully completed. Previously published problems have been automatically republished.").format(course=course_name)}
```
Can we just make this `course_name=course_name`?

potsui replied:
done

stvstnfrd commented on cms/templates/emails/utilities_bulk_update_done_message.txt:
```diff
@@ -0,0 +1,14 @@
+% if success:
+${_("Your bulk update on the {course} course has successfully completed. Previously published problems have been automatically republished.").format(course=course_name)}
+% else:
+${_("Your bulk update on the {course} course has failed due to internal error. Problems in your course may be partially updated. Please contact our team for help.").format(course=course_name)}
```
I suspect `due to internal error` doesn't add much insight to a non-technical user.
Perhaps we can strike it and just say "Your bulk update on the X/Y/Z course has failed."

potsui replied:
done

stvstnfrd commented on cms/templates/emails/utilities_bulk_update_done_message.txt:
```diff
@@ -0,0 +1,14 @@
+${_("Your bulk update on the {course} course has failed due to internal error. Problems in your course may be partially updated. Please contact our team for help.").format(course=course_name)}
+% endif
+
+${_("The following settings were modified:")}
```
It looks like users will see this message even if the import fails.
It seems odd to tell them that we modified settings if there was really a failure.

Should we move this into the `success` block?
Or at the least, we should update the message.

potsui replied:
True, I've moved it to success only

stvstnfrd commented on cms/templates/bulkupdate.html:
```diff
@@ -0,0 +1,133 @@
+  </section>
+</div>
+
+<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.3/jquery.min.js"></script>
```
The platform already ships with its own copy of jquery.
We should probably use that, because this one (1.3) is awfully old anyway.

Grep for `jquery.min.js` for examples.

potsui replied:
done

stvstnfrd commented on cms/templates/bulkupdate.html:
```diff
@@ -0,0 +1,133 @@
+      <div class="bit">
+        <h3 class="title title-3">${_("What is bulk update?")}</h3>
+        <p>
+          ${_("The bulk update page provides a way to update the Max Attempts and Show Answer settings for all existing problems of the current course. This operation also sets the default settings for all future problems of the current course. You can alter this default by going to Advanced Settings.")}
```
We probably don't want to hard code this to the 2 currently supported options,
else this documentation becomes a maintenance burden whenever we add options later.

stvstnfrd replied:
Should this be capitalized, `Bulk Update`?

potsui replied:
done

stvstnfrd commented on cms/templates/bulkupdate.html:
```diff
@@ -0,0 +1,133 @@
+
+    <aside class="content-supplementary" role="complimentary">
+      <div class="bit">
+        <h3 class="title title-3">${_("What is bulk update?")}</h3>
```
Should this be capitalized, `Bulk Update`?

potsui replied:
done

stvstnfrd commented on cms/static/js/views/bulkupdate.js:
```diff
@@ -0,0 +1,341 @@
+        };
+
+        return BulkUpdate;
+    });
```
Can we drop the closing `};` down to a new line,
so that it lines up with the `define(`?

At first it looked like it wasn't being closed.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+        print("Failure sending e-mail for bulk update completion to %s", user_email)
+
+
+class BulkUpdateUtil():
```
Is there any reason for this class with just a single static method?

Can we just make `update_metadata` a top level function?

It can then also probably be "private" (prefix the variable name with an `_`.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+                    try:
+                        value = field.from_json(value)
+                    except Exception as exception:
+                        print(exception)
```
This should be an actual log statement.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+
+                modulestore().update_item(problem, user_id)
+
+                if modulestore().has_published_version(problem):
```
Do we need/want to request the singleton via `modulestore()` each time?
Can we memoize it, ala `store = modulestore(); if store.has_published_version(...)`?

Especially in a loop that could be called hundreds/thousands of times,
those extra calls (inside `modulestore()`') could start to add up...

potsui replied:
no need, replaced with store

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+        _send_email_on_completion(course, user_username, user_email, modified_settings, True)
+    except Exception as exception:
+        print(exception)
+        _send_email_on_completion(course, user_username, user_email, modified_settings, False)
```
We can move this outside the `try/catch`:

```python
success = True
try:
    # pass
except:
    success = False
_send_email_on_completion(course, user_username, user_email, modified_settings, success)
```

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+        course = modulestore().get_course(course_key, 3)
+        _send_email_on_completion(course, user_username, user_email, modified_settings, True)
+    except Exception as exception:
+        print(exception)
```
Remove debug statement.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+
+    try:
+        BulkUpdateUtil.update_metadata(course_key, user_id, modified_settings)
+        course_key = CourseKey.from_string(course_key_string)
```
Why are we calling this again?
The course_key shouldn't change, right?

potsui replied:
Removed. Had forgotten to remove from some earlier debugging

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+    try:
+        BulkUpdateUtil.update_metadata(course_key, user_id, modified_settings)
+        course_key = CourseKey.from_string(course_key_string)
+        course = modulestore().get_course(course_key, 3)
```
Why are we calling this again?

potsui replied:
Removed. Had forgotten to remove from some earlier debugging

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+def bulk_update_problem_settings(course_key_string, user_id, user_username, user_email, modified_settings):
+    course_key = CourseKey.from_string(course_key_string)
+    course = modulestore().get_course(course_key, 3)
+
```
Nit: Remove newline as it doesn't really add clarity here; eg: there aren't newlines around the identical lines below.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+                    except Exception as exception:
+                        print(exception)
+                    field.write_to(problem, value)
+
```
Nit: Remove newlines surrounding `update_item`.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+        """
+        Updates metadata settings for all problems.
+        """
+
```
Elsewhere, there isn't a newline between doctstring and first line of code;
let's remove it here.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+        """
+
+        problems = _get_course_problems(course_key)
+
```
Nit: remove newline.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+            message,
+            from_email,
+            [user_email],
+            fail_silently=False
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+            fail_silently=False
+        )
+    except SMTPException:
+        print("Failure sending e-mail for bulk update completion to %s", user_email)
```
This should be a log statement.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+        'user_username': user_username,
+        'course_name': course.display_name,
+        'modified_settings': modified_settings,
+        'success': success
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+
+    from_email = settings.DEFAULT_FROM_EMAIL
+    subject = render_to_string('emails/utilities_bulk_update_done_subject.txt', context)
+    subject = ''.join(subject.splitlines())
```
Is this because the subject has newlines?

potsui replied:
Yes, or else you will get a `BadHeaderError`

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+
+
+@task()
+def bulk_update_problem_settings(course_key_string, user_id, user_username, user_email, modified_settings):
```
Should we just pass in the `user` object instead of each attribute individually?

potsui replied:
`user` object couldn't be passed as is to a celery task

stvstnfrd replied:
Ah, got it.

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/tasks.py:
```diff
@@ -0,0 +1,97 @@
+    """
+    Retrieve all problems in the course
+    """
+    # Note: Some components may have been orphaned due to a bug in split
```
This could probably be folded into the docstring above.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+            return _utility_bulkupdate_get_handler(request, course_key_string)
+        else:
+            return HttpResponseNotFound()
+
```
Don't place newlines between `if/else` and `try/except` blocks.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+    if not has_course_author_access(request.user, course_key):
+        raise PermissionDenied()
+
+    if 'text/html' in request.META.get('HTTP_ACCEPT', 'text/html'):
```
Why the `in` check instead of `==`?

potsui replied:
The call to `request.META.get('HTTP_ACCEPT', 'text/html')` returns multiple types it can accept as a string, e.g. `'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8'`

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+@login_required
+def utility_bulkupdate_handler(request, course_key_string):
+    """
+    Handler for bulk update requests in the utilities tool. Currently
```
Docstrings are used in a number of "magic" ways in python.
There are some benefits to formatting them like a commit message and/or email:

```
"""
One sentence sound byte

Further explanation here.
"""
```

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+                    'ErrMsg': 'Given value for show answer is not an available option',
+                    'Stage': -1
+                },
+                status=400
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+                'ErrMsg': 'Unable to update advanced settings',
+                'Stage': -2
+            },
+            status=500
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+    # Validate settings
+    _save_request_status(request, session_status_string, 1)
+
+    if max_attempts.find('null') == -1:
```
I'm not familiar with this idiom;
what are you checking?

potsui replied:
When `max_attempts` isn't being changed, its value is passed as `'null'` from the request

stvstnfrd replied:
@potsui So it actually equals `'null'` (the string)?
Should we instead check for that? `if max_attempts != 'null':`?

potsui replied:
good point, done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+    if max_attempts.find('null') == -1:
+        try:
+            max_attempts = int(max_attempts)
+        except:
```
Explicitly catch the specific exception that's thrown,
instead of the blanket except.

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+            return JsonResponse(
+                {
+                    'ErrMsg': 'Given value for max attempts is negative',
+                    'Stage': -1
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+                    'ErrMsg': 'Given value for max attempts is not an integer',
+                    'Stage': -1
+                },
+                status=400
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+                    'ErrMsg': 'Given value for max attempts is negative',
+                    'Stage': -1
+                },
+                status=400
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+            return JsonResponse(
+                {
+                    'ErrMsg': 'Given value for show answer is not an available option',
+                    'Stage': -1
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+            return JsonResponse(
+                {
+                    'ErrMsg': 'Given value for max attempts is not an integer',
+                    'Stage': -1
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+        max_attempts = request.POST.get('maxAttempts')
+        show_answer = request.POST.get('showAnswer')
+        session_status = request.session.setdefault("update_status", {})
+        session_status_string = course_key_string + max_attempts + show_answer
```
You probably need a delimiter in there.

Consider:
```python
'Stanford/Course/Fall2017` + 10 + True
```

vs

```python
'Stanford/Course/Fall20171` + 0 + True
```

potsui replied:
How about this? 

`return 'course_key_string={}&max_attempts={}&show_answer={}'.format(
        course_key_string,
        max_attempts,
        show_answer,
    )`

stvstnfrd commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        {'showAnswer': 'invalid_show_answer_option'},
+        {'maxAttempts': -1, 'showAnswer': SHOW_ANSWER_OPTIONS[0]},
+        {'maxAttempts': 0, 'showAnswer': 'invalid_show_answer_option'},
+        {'maxAttempts': -1, 'showAnswer': 'invalid_show_answer_option'}
```
https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

potsui replied:
done

stvstnfrd commented:
Oh, and tests are failing :)

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+"""
```
Done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+"""
```
Done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        response = self.client.get(self.bulkupdate_url, HTTP_ACCEPT='text/html')
+        self.assertContains(response, "Update Problem Settings")
+        # The HTML generated will define the handler URL (for use by the Backbone model).
+        self.assertContains(response, self.bulkupdate_url)
```
Removed

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+
+    @ddt.data(
+        {'maxAttempts': 0, 'showAnswer': SHOW_ANSWER_OPTIONS[0]},
+        {'maxAttempts': 1, 'showAnswer': SHOW_ANSWER_OPTIONS[1]}
```
Done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+
+    @ddt.data(
+        {'maxAttempts': 0},
+        {'showAnswer': SHOW_ANSWER_OPTIONS[0]}
```
done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        Tests POST operation works given two correct arguments
+        """
+        update_url = reverse_course_url('utility_bulkupdate_handler', self.course.id)
+        print settings
```
done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        """
+        Tests POST operations works given one correct argument
+        """
+        update_url = reverse_course_url('utility_bulkupdate_handler', self.course.id)
```
done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        Tests POST operations works given one correct argument
+        """
+        update_url = reverse_course_url('utility_bulkupdate_handler', self.course.id)
+        print settings
```
done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        Tests POST operation returns 'invalid setting' 400 code on incorrect arguments
+        """
+        update_url = reverse_course_url('utility_bulkupdate_handler', self.course.id)
+        print settings
```
done

potsui commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+    'finished',
+    'past_due',
+    'correct_or_past_due',
+    'never'
```
done

potsui commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+                setattr(course, key, value)
+
+            modulestore().update_item(course, request.user.id)
+
```
done

potsui commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+            'bulkupdate_status_url': status_url,
+            'default_max_attempts': DEFAULT_MAX_ATTEMPTS,
+            'default_show_answer': DEFAULT_SHOW_ANSWER,
+            'show_answer_options': SHOW_ANSWER_OPTIONS
```
done

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        Tests getting the HTML template for the bulkupdate page
+        """
+        response = self.client.get(self.bulkupdate_url, HTTP_ACCEPT='text/html')
+        self.assertContains(response, "Update Problem Settings")
```
Removed

potsui commented on cms/djangoapps/contentstore/views/__init__.py:
```diff
@@ -8,6 +8,7 @@
 from .utility import *
 from .utilities.captions import *
 from .utilities.bulksettings import *
+from .utilities.bulkupdate import *
```
done

potsui commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+"""
+Views related to bulk settings change operations on course problems.
```
dones

potsui commented on cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,245 @@
+
+from opaque_keys.edx.keys import CourseKey
+
+from contentstore.views.utilities.tasks import bulk_update_problem_settings
```
done

potsui commented:
@stvstnfrd @caseylitton pushed a WIP commit addressing feedback so far. Casey and I have been working on moving files over to `openedx` but running into a few issues with URL matching

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        Creates the test course
+        """
+        super(BulkUpdateTest, self).setUp()
+        self.course = CourseFactory.create(org='edX', number='test1', display_name='BulkUpdate Course')
```
Done! Also, my issue in writing unit tests right now is related to this. The `course.id` being returned is slash separated which is deprecated; I'm still looking through documentation to find a way to convert but haven't found anything yet @stvstnfrd

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+
+    def test_get_bulkupdate(self):
+        """
+        Tests the get bulkupdate method and URL expansion
```
changed

potsui commented on cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,124 @@
+        self.assertContains(response, self.bulkupdate_url)
+        # Verify dynamic content populates HTML correctly
+        self.assertContains(response, '<input class="input setting-input setting-input-number" type="number" value="0" min="0.0000" step="1">')
+        self.assertContains(response, '<option value="always" selected>always</option>')
```
done

potsui commented on openedx/stanford/cms/envs/common.py:
```diff
@@ -32,6 +32,16 @@
                 'action_external': False,
             },
+            {
+                'short_description': 'Bulk update problem settings',
```
Might as well? I was following the other utilities already listed; should I change those too?

potsui commented on cms/urls.py:
```diff
@@ -85,6 +85,8 @@
     url(r'^utilities/{}$'.format(settings.COURSE_KEY_PATTERN), 'utility_handler'),
     url(r'^utility/captions/{}$'.format(settings.COURSE_KEY_PATTERN), 'utility_captions_handler'),
     url(r'^utility/bulksettings/{}$'.format(settings.COURSE_KEY_PATTERN), 'utility_bulksettings_handler'),
+    url(r'^utility/bulkupdate/{}$'.format(settings.COURSE_KEY_PATTERN), 'utility_bulkupdate_handler'),
```
Was having issues with this. Per f2f, reverted this change for the time being.

potsui commented:
@stv @casey I've finished going through all the feedback. Will clean up a bit more and make a new PR for easier review.
